registry/storage/driver/s3-aws/s3.go

```diff
@@ -498,6 +519,58 @@ func New(params DriverParameters) (*Driver, error) {
 	}, nil
 }
 
+// logS3OperationHandlerName is used to identify the handler used to log S3 API
+// requests
+const logS3OperationHandlerName = "docker.storage-driver.s3.operation-logger"
+
+// logS3Operation logs each S3 operation, including request and response info,
+// as it completes
+func logS3Operation(ctx context.Context) request.NamedHandler {
+	return request.NamedHandler{
+		Name: logS3OperationHandlerName,
+		Fn: func(r *request.Request) {
+			req := r.HTTPRequest
+			resp := r.HTTPResponse
+			op := r.Operation
+			fields := map[interface{}]interface{}{
+				"s3_operation_name":                        op.Name,
+				"s3_http_request_method":                   req.Method,
+				"s3_http_request_host":                     req.Host,
+				"s3_http_request_path":                     req.URL.Path,
+				"s3_http_request_content-length":           req.ContentLength,
+				"s3_http_request_remote-addr":              req.RemoteAddr,
+				"s3_http_response_header_x-amz-request-id": resp.Header.Values("x-amz-request-id"),
+				"s3_http_response_status":                  resp.StatusCode,
+				"s3_http_response_content-length":          resp.ContentLength,
+			}
+
+			// X-Do-Spaces-Error is a header set on HTTP responses from Digtalocean's
+			// Spaces S3 api that provides additional context on api errors when they
+			// occur. checking its length here before adding it to the log fields to
+			// avoid confusing users of other S3 APIs.
+			doSpacesError := resp.Header.Get("x-do-spaces-error")
```

Comment:
Rather than hard-coding this, could we accept a list of headers to log as a configuration parameter? That would let users of any S3-compatible object storage configure their own custom headers for logging.

Reply:
Done, thanks for the suggestion!